mesonbuild/dependencies/pkgconfig.py

```diff
@@ -0,0 +1,327 @@
+                else:
+                    # '\?' - Include both chars.
+                    start = i - 1
+            elif c == '\\':
+                # Next char is escaped, copy everything we had so far.
+                escaped = True
+                line += content[start:i]
+            elif c == '#':
+                # '#' ends the current line. Next line will start after the
+                # first '\n' we'll find.
+                comment = True
+                line += content[start:i]
+                yield line
+                line = ''
+            elif c == '\n':
+                # '\n' ends the current line. Next like starts at the next char.
+                line += content[start:i]
+                yield line
+                line = ''
+                start = i + 1
+
+    def _get_args(self, args, system_args, allow_system_args):
+        if allow_system_args:
+            return args.copy()
+        return [a for a in args if a not in system_args]
+
+    def get_cflags(self, allow_system_cflags=False):
+        cflags = self._get_args(self.cflags, self.repo.system_cflags, allow_system_cflags)
+        for pkg in self.requires_private:
+            cflags.extend(pkg.get_cflags(allow_system_cflags))
+        for pkg in self.requires:
+            cflags.extend(pkg.get_cflags(allow_system_cflags))
+        return cflags
+
+    def get_libs(self, static, allow_system_libs=False):
+        libs = self._get_args(self.libs, self.repo.system_libs, allow_system_libs)
+        for pkg in self.requires:
+            libs.extend(pkg.get_libs(static))
+        if static:
+            libs.extend(self._get_args(self.libs_private, self.repo.system_libs, allow_system_libs))
+            for pkg in self.requires_private:
+                libs.extend(pkg.get_libs(static))
+        return libs
+
+    def get_variable(self, varname, default=None, definition=None):
+        if not definition:
+            return self.variables.get(varname, default)
+
+        defname, defvalue = definition
+        if varname == defname:
+            return defvalue
+
+        # Reevaluate all variables in order, with the new definition, until we
+        # find the one we want.
+        variables = {defname: defvalue}
+        for key, raw_value in self.raw_variables.items():
+            if key == defname:
+                continue
+            value = self._expand(raw_value, variables)
+            if key == varname:
+                return value
+            variables[key] = value
+
+        if default is not None:
+            return default
+
+        raise PkgConfigException('Unknown variable name: {!r}'.format(varname))
+
+class Repository:
+    def __init__(self, system_paths=None, extra_paths=None, sysroot_dir=None, sysroot_map=None):
+        system_prefix = 'c:/' if mesonlib.is_windows() else '/usr'
+        self.system_libdir = Path(system_prefix, mesonlib.default_libdir())
+        self.system_incdir = Path(system_prefix, 'include')
```

Comment:
FWIW, I think that's the problematic part. Those values are typically configured by distro when they build pkg-config package.

OTOH, we already do distro specific value for `mesonlib.default_libdir()`, but we probably would need to make that configurable by distro. I don't know how packaging works for python projects, is there some standard way of doing packaging configuration?

Reply:
There is not, and you can't make a pip-installable package that has different bytes of any sort on, say, Debian vs Fedora. The only point of configurability is the operating system (Windows, macOS, or Linux) and which libc a compiled C extension is compatible with (manylinux or musllinux).

Wheels are distribution-independent, relocatable, zero-configuration and zero-flexibility analogues to .deb files with the proviso that there's no actual database.

If we allow people to get meson from PyPI by using pip install, we automatically cannot under any circumstances support configuring that prebuilt package, and we can't really support building from source either because users will not know how to manually pass options to pip that are read by meson (and they shouldn't have to, because the distro already correctly configured pkg-config).

The even bigger issue is what to do when the user modifies PATH to add an alternative pkg-config implementation and an alternative gcc at the front, in order to build against a non-default SDK. This is a breaking change.

Reply:
> The even bigger issue is what to do when the user modifies PATH to add an alternative pkg-config implementation and an alternative gcc at the front, in order to build against a non-default SDK. This is a breaking change.

A different pkg-config in PATH is indeed a problem we cannot really detect. That would be the same issue if we use libpkgconfig, I think we should just accept it's not supported any more at some point. However, we can keep using the CLI if PKG_CONFIG is set, my initial plan was to use the internal implementation when `PKG_CONFIG=meson` at the beginning, as experiment.

For the configuration I found that we can get builtin search paths using `pkg-config --variable pc_path pkg-config`. That works with both pkg-config and pkgconf. Unfortunately getting system include and libraries dirs seems only supported by pkgconf using `pkg-config --variable pc_system_libdirs pkg-config` and `pkg-config --variable pc_system_includedirs pkg-config`.

On systems where pkg-config or pkgconf are not installed, we could just use Meson defaults, that would not be a regression at least.